p2p/chain.py

```diff
@@ -720,6 +610,138 @@ class RegularChainSyncer(FastChainSyncer):
+                token=self.cancel_token)
+            uncles = await wait_with_token(
+                chaindb.coro_get_block_uncles(header.uncles_hash),
+                token=self.cancel_token)
+            bodies.append(BlockBody(transactions, uncles))
+        peer.sub_proto.send_block_bodies(bodies)
+
+    async def handle_get_receipts(self, peer: ETHPeer, block_hashes: List[Hash32]) -> None:
+        chaindb = cast('AsyncChainDB', self.db)
+        receipts = []
+        # Only serve up to eth.MAX_RECEIPTS_FETCH items in every request.
+        for block_hash in block_hashes[:eth.MAX_RECEIPTS_FETCH]:
+            try:
+                header = await wait_with_token(
+                    chaindb.coro_get_block_header_by_hash(block_hash),
+                    token=self.cancel_token)
+            except HeaderNotFound:
+                self.logger.debug(
+                    "%s asked receipts for block we don't have: %s", peer, block_hash)
+                continue
+            block_receipts = await wait_with_token(
+                chaindb.coro_get_receipts(header, Receipt),
+                token=self.cancel_token)
+            receipts.append(block_receipts)
+        peer.sub_proto.send_receipts(receipts)
+
+    async def handle_get_node_data(self, peer: ETHPeer, node_hashes: List[Hash32]) -> None:
+        chaindb = cast('AsyncChainDB', self.db)
+        nodes = []
+        # Only serve up to eth.MAX_STATE_FETCH items in every request.
+        for node_hash in node_hashes[:eth.MAX_STATE_FETCH]:
+            try:
+                node = await wait_with_token(
+                    chaindb.coro_get(node_hash),
+                    token=self.cancel_token)
+            except KeyError:
+                self.logger.debug("%s asked for a trie node we don't have: %s", peer, node_hash)
+                continue
+            nodes.append(node)
+        peer.sub_proto.send_node_data(nodes)
+
+    async def lookup_headers(self, block_number_or_hash: Union[int, bytes], max_headers: int,
+                             skip: int, reverse: bool) -> List[BlockHeader]:
+        """
+        Lookup :max_headers: headers starting at :block_number_or_hash:, skipping :skip: items
+        between each, in reverse order if :reverse: is True.
+        """
+        try:
+            block_numbers = await self._get_block_numbers_for_request(
+                block_number_or_hash, max_headers, skip, reverse)
+        except HeaderNotFound:
+            self.logger.debug(
+                "Peer requested starting header %r that is unavailable, returning nothing",
+                block_number_or_hash)
+            block_numbers = []
+
+        headers = [header async for header in self._generate_available_headers(block_numbers)]
+        return headers
+
+    async def _get_block_numbers_for_request(
+            self, block_number_or_hash: Union[int, bytes], max_headers: int,
+            skip: int, reverse: bool) -> List[BlockNumber]:
```

Comment:
Mostly unrelated.  I think go-ethereum recently patched a bug which exploited overflowing their integer with very large `skip` values.  While we don't suffer from the same overflow problems, it makes me think that we should enforce a reasonable upper bound on the skip size.